### PR TITLE
Bitmasked bitshifting

### PR DESCRIPTION
**Pull request by krux02**
This is a breaking change that does not yet have a switch to enable old behavior, so don't merge it yet.
The goal of this change is to evade undefined behavior without sacrificing performance.

This PR changes (again) how bitshifting works. Bitshifting is compiled to the C bitshiftoperator, but here it is undefined behavior when the operand is bigger than the amount of bits in the bitshifted value. To fix this, in the past all bit operations were done in ``BiggestInt``, which didn't fix the problem at all, because undefenied behavior was still triggered for operands from 64 and upwards.

In this PR all bit shifting operations mask their argument so that it will never exceed the capacity of the interger type. I don't think this behavior is particularly useful, but it evades undefined behavior nicely and it is fast.

### CODE REVIEW

File: tests/arithm/tarithm.nim [hunk 1]
```diff
@@ -135,9 +154,9 @@ block tshl:
     const t1: uint16 = 1'u16 shl 16
     const t2: uint32 = 1'u32 shl 32
     const t3: uint64 = 1'u64 shl 64
-    doAssert t0 == 0
-    doAssert t1 == 0
-    doAssert t2 == 0
+    doAssert t0 == 1
+    doAssert t1 == 1
+    doAssert t2 == 1
```
Comment (mratsim) [inline tests/arithm/tarithm.nim]: We should be consistent with C as mentioned in https://github.com/nim-lang/Nim/issues/11138

For example uint32 numbers are implicitly modulo 2^32
`2^32 mod 2^32 == 0` not 1

This makes Nim VM unusable for big int libraries like [stint](https://github.com/status-im/nim-stint) or [constantine](https://github.com/status-im/nim-constantine)

Comment (krux02) [reply tests/arithm/tarithm.nim]: You misunderstand. The right operand gets the modulo operation. ``1'u32 shl 32`` = ``1'u32 shl (32 mod 32)`` = ``1'u32 shl 0`` = ``1'u32``. 

Here is a motivation for this change from [cppreference.com](https://en.cppreference.com/w/cpp/language/operator_arithmetic)

> In any case, if the value of the right operand is negative or is greater or equal to the number of bits in the promoted left operand, the behavior is undefined.

In other words ``1'u64 shl 64`` was mapped to undefined behavior in c/c++.

**Comment by arnetheduck** [comment]
this looks quite odd and unintuitive though...
```
var x = 42'u64
for i in 0..<64:
  x = x shr 1
```
and
```
x shr 64
```
feel like they should give the same result..

**Comment by krux02** [comment]
@arnetheduck The bitmasking is applied, because that is what the assempler instructions ``sar, sal, shr, shl`` do as well. The explicit bitmaking is done, so that this behavior is punched through the C compilation stage as well, because here it is undefined behavior if the argument is too big. 

Undefined really means undefined, giving different results on different platforms:
https://blogs.msdn.microsoft.com/larryosterman/2011/02/11/the-case-of-the-inconsistent-right-shift-results/

**Comment by narimiran** [comment]
@krux02 What is the status of this one? Do you plan to continue working on it or should it be closed?

**Comment by krux02** [comment]
Don't close it yet. I go regularly through my inactive pull requests to see if they are still valuable. This one is even though non-merged. We should have this auto closing bot though that sends reminders to inactive PRs.

**Comment by ridho9** [comment]
What is the current status of this?

**Comment by Araq** [comment]
@ridho9 We will merge this once the CIs with old GCC versions don't crash on the produced code.

**Comment by mratsim** [comment]
One question, is there an escape hatch for a library that knows that the arguments do not overflow the shifts?

For example a crypto library where every single operation count. All Ethereum 2 clients spend 20%~50% of their time in crypto so it's very important that it's as fast as possible.

**Comment by arnetheduck** [comment]
fwiw, python works as one would (naturally) expect it to:

```
Python 3.7.4 (default, Jul  9 2019, 16:32:37) 
[GCC 9.1.1 20190503 (Red Hat 9.1.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 1 >> 64
0
```

**Comment by Araq** [comment]
Huh? what? I would have expected Python to use its bignums to produce 2^64!

**Comment by narimiran** [comment]
> I would have expected Python to use its bignums to produce 2^64!

Note that @arnetheduck used right-shift (`>>`) in his example, not left-shift.

```python
In [1]: 1 >> 64                                                                                                   
Out[1]: 0

In [2]: 1 << 64                                                                                                   
Out[2]: 18446744073709551616
```

**Comment by mattrbeck** [comment]
Just to throw another opinion in, I personally favor defining the result of out-of-range bitshifts to be 0. For reference (although it doesn't need to dictate the decision here), Crystal treats those as 0: [link](https://github.com/crystal-lang/crystal/blob/dd40a2442fa186add8a82b74edb14a90aa1dae05/src/int.cr#L203). Either way, I believe Nim _should_ define this undefined C behavior so that it doesn't vary across systems.

**Pull request closed without merging**